File: DSA/src/Questions/BinaryTree/KthSmallestElement.java
package Questions.BinaryTree;

public class KthSmallestElement {
    public int inOrder(TreeNode root, int[] k) {
        if (root == null)
            return -1;
        int left = inOrder(root.left, k);
        k[0]--;
        if (k[0] == 0)
            return root.val;
        int right = inOrder(root.right, k);
        if (left != -1) {
            return left;
        } else {
            return right;
        }

    }

    public int kthSmallest(TreeNode root, int k) {
        return inOrder(root, new int[] { k });
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        TreeNode left = new TreeNode(1);
        TreeNode right = new TreeNode(4);
        TreeNode leftRight = new TreeNode(2);
        root.left = left;
        root.right = right;
        left.right = leftRight;
        KthSmallestElement kthSmallestElement = new KthSmallestElement();
        int result = kthSmallestElement.kthSmallest(root, 1);
        System.out.println("Result: " + result);
    }
}

File: DSA/Python/Questions/BinaryTree/KthSmallestElement.py
"""
# Question: Kth Smallest Element in a BST
# Link: https://leetcode.com/problems/kth-smallest-element-in-a-bst/

# Time Complexity: O(H + k)
# Space Complexity: O(H)

# Algorithm:
# 1. Inorder traversal
# 2. Track count
# 3. Return kth element
"""


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


class KthSmallestElement:
    def kth_smallest(self, root: TreeNode, k: int) -> int:
        self.k = k
        self.result = 0

        def inorder(node: TreeNode) -> None:
            if not node:
                return

            inorder(node.left)
            self.k -= 1
            if self.k == 0:
                self.result = node.val
                return
            inorder(node.right)

        inorder(root)
        return self.result


def main():
    solution = KthSmallestElement()
    root = TreeNode(3)
    root.left = TreeNode(1)
    root.right = TreeNode(4)
    print(solution.kth_smallest(root, 1))


if __name__ == "__main__":
    main()
